sdlib/d/gc/extent.d

```diff
@@ -306,57 +306,127 @@ public:
 	}
 
 	void setFreeSpace(uint index, size_t freeSpace) {
 		assert(isSlab(), "setFreeSpace accessed on non slab!");
-		assert(freeSpace <= slotSize, "freeSpace exceeds alloc size!");
+		assert(freeSpace <= getTotalCapacity(index),
+		       "freeSpace exceeds alloc size!");
 		assert(index < slotCount, "index is out of range!");
-		assert(supportsFreeSpace, "size class not supports freeSpace!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
-		if (freeSpace == 0) {
-			freeSpaceFlags.clearBitAtomic(index);
+		// If we have a finalizer, the freeSpace may be 0:
+		if (freeSpace == 0 && !hasFinalizer(index)) {
+			slabMetaDataFlags.clearBitAtomic(index);
 			return;
 		}
 
 		// Encode freespace and write it to the last byte (or two bytes) of alloc.
 		writePackedFreeSpace(freeSpacePtr(index), freeSpace & ushort.max);
-		freeSpaceFlags.setBitAtomic(index);
+		slabMetaDataFlags.setBitAtomic(index);
 	}
 
 	size_t getFreeSpace(uint index) {
 		assert(isSlab(), "getFreeSpace accessed on non slab!");
 		assert(index < slotCount, "index is out of range!");
 
-		if (!supportsFreeSpace || !freeSpaceFlags.valueAtAtomic(index)) {
+		if (!hasSlabMetaData(index)) {
 			return 0;
 		}
 
 		// Decode freespace, found in the final byte (or two bytes) of the alloc:
 		return readPackedFreeSpace(freeSpacePtr(index));
 	}
 
+	size_t getTotalCapacity(uint index) {
+		assert(isSlab(), "getTotalCapacity accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		if (!hasFinalizer(index)) {
+			return slotSize;
+		}
+
+		return slotSize - PointerSize;
+	}
+
+	enum FinalizerBit = nativeToBigEndian!ulong(ulong(0x2));
+
+	bool hasFinalizer(uint index) {
+		assert(isSlab(), "hasFinalizer accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		return hasSlabMetaData(index)
+			&& (*(finalizerPtr(index)) & FinalizerBit);
+	}
+
+	ulong* finalizerPtr(uint index) {
+		assert(isSlab(), "finalizerPtr accessed on non slab!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
+
+		return cast(ulong*) (address + (index + 1) * slotSize) - 8;
+	}
+
+	Finalizer getFinalizer(uint index) {
+		assert(isSlab(), "getFinalizer accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		if (!hasSlabMetaData(index)) {
+			return null;
+		}
+
+		auto finalizerFieldPtr = finalizerPtr(index);
+		if (!(*finalizerFieldPtr & FinalizerBit)) {
+			return null;
+		}
+
+		return cast(Finalizer) cast(void*) (*finalizerFieldPtr & AddressMask);
+	}
+
+	void setFinalizer(uint index, Finalizer finalizer) {
+		assert(isSlab(), "setFinalizer accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+		assert(hasSlabMetaData(index),
+		       "No metadata! (must set freespace before finalizer)");
+
+		auto finalizerFieldPtr = finalizerPtr(index);
+
+		if (finalizer is null) {
+			*finalizerFieldPtr &= ~FinalizerBit;
```

Comment:
If freeSpace is zero, don't you want to clear the metadata bit? On the other hand,c an freeSpace be zero in the presence of a finalizer? Probably not, so we might want to add the appropriate assertions that this is the case.